Stabilize Histogram to Prometheus translation and specify NHCB conversion (#5091)
open-telemetry/opentelemetry-specification: specification/compatibility/prometheus_and_openmetrics.md

Reviewer: This section already describes how to convert otel exponential histograms to prometheus native histograms. If you want to expose your histogram instruments as prometheus native histograms, why not just configure either:

- The entire prometheus reader to configure base2 exponential histograms as the aggregation for all histogram type metrics https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/sdk.md#L1346C9-L1346C9
- A view to select that specific histogram instrument should be aggregated to base2 exponential histogram

It seems needlessly complicated to add an optional configuration param to convert explicit bucket histograms to prometheus native histograms when we already have tools available to configure the SDK to aggregate data to otel exponential histograms, which are a much closer match.

Reviewer 2: This is an unfortunate artifact of the internal storage of prometheus leaking into Protocols. See https://github.com/open-telemetry/opentelemetry-specification/pull/5091#issuecomment-4445095729.

As a performance optimization, Prometheus can store classic (fixed-bucket) histograms as native (exponential) histograms in its TSDB, with a special `schema=-53`, and refers to these as "Native Histograms with custom buckets", or NHCBs.  In Prometheus remote write 2.0, you can choose to send a fixed-bucket histogram as either this NHCB, or as individual float series (with _sum, _count, _bucket suffixes). Now that this concept is in protocols, it feels like we have to define how to deal with them. I'm hoping to put the cat back in the bag, but I'm not sure that will be possible.

At least this won't apply to SDK exporters, but maybe we need to make that clearer.

Reviewer: > At least this won't apply to SDK exporters, but maybe we need to make that clearer.

👌

Reviewer 2: This section now says:

> NHCB output is currently only supported by the [Prometheus Remote-Write 2.0
or later](https://prometheus.io/docs/specs/prw/remote_write_spec_2_0/) protocol.

So hopefully that is clear enough

### Histograms

**Status**: [Stable](../document-status.md)

> An [OpenTelemetry Histogram](../metrics/data-model.md#histogram) with a cumulative aggregation temporality MUST be converted to a Prometheus Histogram by default, but users may opt-in to converting an OpenTelemetry Histogram to a Prometheus Native histogram with custom buckets (NHCB) for protocols that support NHCB.

Author: After talking to @ArthurSens : I'll make this more specific and say that the optionality only concerns cases when we convert to Prometheus Remote-Write. As otherwise no Prometheus exposition format supports NHCB directly.

When converting to a Prometheus Histogram, the following metrics MUST be created:

> - A single `{name}_count` metric denoting the count field of the histogram. All attributes of the histogram point are converted to Prometheus labels.

Reviewer 2: Should we make this more agnostic to the input format?  E.g. just talk about "count" and "sum", etc. instead of _count series, _sum series?

Author: That's the output format and I don't think we have a choice, we need to say that the name of the metric must be,.

Reviewer 2: But that isn't always what the metric name will be, right?

Author: I don't know what you mean.

Reviewer 2: At least in OM 2.0 we won't have a _count metric, for example.

Author: As discussed I've rewritten this to be model to model instead of model to exposition style in ac6a9519c55ecddb8350018519663d6e050d5930 .

Two major changes:

### Metric names

Metric names are no longer explicitly defined. Left to the client library. If we wanted to describe them, we'd need to say something like:

If the Prometheus Protocol requires supplying the metric names for encoding
the `Count`, `Sum` and `Buckets`, then

- All attributes of the histogram point are converted to Prometheus labels
  for all resulting metrics.
- The `Count` MUST correspond to a single `{name}_count` metric denoting
  the count field of the histogram.
- The `Sum` MUST correspond to a single `{name}_sum` metric denoting the
  sum field of the histogram.
- The `Buckets` correspond to a series of `{name}_bucket` metrics denoting
  the cumulative count of each bucket. For each bucket, the upper bound is
  denoted as `le` label which is added to the set of labels. The `le`
  label's value is the stringified floating point value of bucket
  boundaries, ordered from lowest to highest. The final metric MUST have
  the `le` equal `+Inf` label.

### Sum not optional anymore

Removes "reported only if the sum is positive and monotonic". There were a couple things wrong with this. Not all client libraries do this distinction (e.g. client_golang doesn't care at all). Not all formats would even require dropping sum (e.g. protobuf/openmetrics 2.0).  Also it should have read "non-negative" or positive or zero".

Reviewer 2: lgtm

- `{name}_sum` metric denoting the sum field of the histogram, reported only if the sum is positive and monotonic. The sum is positive and monotonic when all buckets are positive. All attributes of the histogram point are converted to Prometheus labels.
- A series of `{name}_bucket` metric points that contain all attributes of the histogram point recorded as labels.  Additionally, a label, denoted as `le` is added denoting the bucket boundary. The label's value is the stringified floating point value of bucket boundaries, ordered from lowest to highest. The value of each point is the sum of the count of all histogram buckets up to the boundary reported in the `le` label.  The final bucket metric MUST have an `+Inf` threshold.
- `Min` and `Max` are not used.
- `Exemplars` are converted as described in the [Exemplar Conversion](#exemplar-conversion) section. If the Prometheus protocol only supports a single exemplar per-bucket, the latest exemplar that falls into each bucket SHOULD be converted.
- If set, `StartTimeUnixNano` SHOULD be transformed into Prometheus `StartTime`,
  following the appropriate format used by each Prometheus protocol.

> When converting to a Prometheus NHCB, only a single NHCB metric MUST be created:

Reviewer 2: Please keep this section in development.  It won't be required for SDKs, so I think thats fine for now.

Author: Do you mean NHCB specifically or the whole histogram section ?

Reviewer 2: NHCB specifically

Reviewer 3: Adding a whole new section and calling it "Stable" already sounds like a bold move. I agree that we can introduce this in development. We could probably break this PR down into two:
1) One to add the NHCB bits
2) Stabilize the Classic Histogram parts

Reviewer 3: Splitting the PRs will also allow us to move forward with SDK work because NHCB isn't relevant to SDKs at all

Author: I've updated the PR. It should be fairly easy to split this PR into two now, but that would mean touching the "stable" bit, so I'd massage this PR first a little bit to see if we can merge it in one piece.

- The [flavor](https://prometheus.io/docs/specs/native_histograms/#flavors)
  of the NHCB MUST be integer counter.

> - The `ResetHint` in the NHCB MUST be set to `UNKNOWN`. OpenTelemetry does not

Author: Note to reviewers: this has been discussed before, hence:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b35feafebc4f952a99c944a3daaed3a96d2692ab/pkg/translator/prometheusremotewrite/histograms_v2.go#L73

Reviewer 4: My understanding of OTel's "Reset signal" is that it is implied when `StartTimestamp==Timestamp` for any data point.

This is written about in the data model specification: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#cumulative-streams-inserting-true-reset-points

According to that, you could decide to consider only true resets, those being the case where `StartTimestamp == Timestamp && point.IsZero()` (i.e., count/sum are both zero).

In my understanding, the other case, "reset following gap" i.e., `StartTimestamp == Timestamp && !point.IsZero()`, or "observed reset", these are what we get when scraping a target for the first time.

@krajorama I would appreciate your help in clarifying what OpenTelemetry can do to better support Prometheus on this topic, thanks!

Reviewer 2: The prometheus counter reset field is a performance optimization to avoid having to do reset detection again. The main issue (from my understanding, @krajorama might be able to provide more detail), is that this field is very tightly coupled to logic in the Prometheus server's TSDB.  See https://prometheus.io/docs/specs/native_histograms/#counter-reset-considerations for all of the logic that goes into properly setting this field. Since UNKNOWN is always safe, it seems like the right behavior for us, at least for now.

I could see an argument for making this SHOULD so that we leave the door open to doing this in the future.

Reviewer 3: I wouldn't be against a SHOULD here and leaving this open for folks who know what they are doing.

Author: Our current plan for taking into account the StartTimestamp is to either inject a zero sample (feature [created-timestamp-zero-injection](https://prometheus.io/docs/prometheus/latest/feature_flags/#start-created-timestamps-zero-injection)) or store the StartTimestamp per sample and take it into account in queries (feature [st-storage](https://prometheus.io/docs/prometheus/latest/feature_flags/#start-timestamp-st-native-storage) and [use-start-timestamps](https://prometheus.io/docs/prometheus/latest/feature_flags/#start-timestamp-st-usage-in-promql-functions)), see also https://github.com/prometheus/prometheus/pull/18619. Neither require recording the native histogram sample with explicit reset hint based on the StartTimestamp.

So from correctness point of view saying MUST UNKNOWN will be ok.

From optimization point of view it would be nice to say SHOULD for the specific case of StartTimestamp==[End]Timestamp, since it avoids Prometheus having to recalculate the reset - if and only if StartTimestamp==[End]Timestamp is ensured 100% to mean a counter reset before the sample. However that's not what the spec says, it says "When StartTimeUnixNano equals TimeUnixNano, a new unbroken sequence of observations begins with a reset at an unknown start time. The initial observed value is recorded to indicate that an unbroken sequence of observations resumes. These points have zero duration, and indicate that **nothing is known about previously-reported points** and that data may have been lost." Nothing is known about previously-reported points to me means we don't know if there's a reset -> hence the trick with replacing the first value with zero and subtracting it from subsequent values as [Cumulative streams: inserting true reset points](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#cumulative-streams-inserting-true-reset-points) describes.

So I think I'd stick with UNKNOWN.

Reviewer 4: Can we keep the "SHOULD" in that case? My recommendation was that RESET points can be inferred when both the start timestamps are equal _and the value is zero_, which I think is what you're saying.

Reviewer 5: I agree with @krajorama. From my understanding of OTel spec, `ST = T` means unknown start time, so it follows to set UNKNOWN reset hint. 

Though I think I also see the angle where @jmacd is coming from. `ST = T` datapoint might contain enough information to be able to argue that other reset hint types could be used without impacting `rate()` correctness. For example, if a datapoint has `ST = T`, it means that it's the first datapoint in a sequence of scrapes. And if it's also `point.IsZero()`, then you technically wouldn't _break_ anything by setting reset hint to YES, because under normal circustances `rate()` result would be the same as with UNKNOWN. However, YES wouldn't be factually true,  because `ST == T && point.IsZero()` datapoint might be a continuation after a gap of a counter, which previously also had zero values (i.e. there was no reset, but there was a gap because of which the start time is unknown).

Either way, I think the most natural choice is to use UNKNOWN reset hint. But as others mentioned, making it SHOULD would allow optimization which do not impact `rate()` results, e.g. as described above.

  carry an explicit reset flag, so `UNKNOWN` lets Prometheus auto-detect resets
  from the values and avoids any semantic divergence with OpenTelemetry's reset
  model.
- The `Schema` in the NHCB MUST be set to -53.

> - `Count` is converted to Native Histogram `Count` if the `NoRecordedValue`

Author: note to reviewers: integer native histogram cannot carry the staleness marker in the Count as the staleness marker is a specific 64bit pattern relevant for floats. Also Native histograms will ignore every other field if sum is equal to the staleness marker. See https://prometheus.io/docs/specs/native_histograms/#staleness-markers

I'll fix this in the exponential histograms conversion section below, when stabilizing that.

Reviewer 3: Could you add this link to the specification itself? I think it will help future readers understand the decision made here.

Author: Yes. But I'm adding to the `Sum` discussion and removing setting to zero, that's the default value and I don't think we need to specify for every single field individually that it should be set to zero if not used.

Also I just realized that in the [Otel model](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#histogram) the `Sum` is not optional, so there's no point in saying "if it is set".

Author: Actually I'll follow David's suggestion in https://github.com/open-telemetry/opentelemetry-specification/pull/5091#discussion_r3237466137

> - `Count` is converted to Native Histogram `Count` if the `NoRecordedValue`

Reviewer 2: Would it be better to just have a small subsection dedicated to what to do when NoRecordedValue is true? E.g. should we still convert bucket counts if the NoRecordedValue is set?

Author: done

- If set, `StartTimeUnixNano` SHOULD be transformed into Prometheus `StartTime`,
  following the appropriate format used by each Prometheus protocol.
- The bucket boundaries are written into the Native Histogram `CustomValues` in
  ascending order. The implicit `+Inf` upper bound of the final OTel bucket MUST
  NOT be written into `CustomValues`; it is represented by the overflow bucket
  at index `len(CustomValues)`.
- The dense `BucketCounts` are converted into the [sparse bucket layout](https://prometheus.io/docs/specs/native_histograms/#buckets)
  in `PositiveSpans` and `PositiveDeltas`.
  - Non-zero bucket counts MUST be encoded into `PositiveDeltas`. Zero-count
    buckets MAY be encoded (as a zero delta inside an enclosing span) to reduce
    the number of `PositiveSpans`; otherwise they appear as gaps between spans.
  - Note: buckets with negative boundary are also encoded into `PositiveSpans`
    and `PositiveDeltas`.
  - Bucket counts that need to be converted are converted into `PositiveDeltas`,
    which is delta encoded. The first converted value is written as is, the rest
    as delta to the previous value. Unlike the classic histogram conversion

>     above, no cumulative summation across buckets is required — OTel

Reviewer 2: ```suggestion
    above, no cumulative summation across buckets is required — OpenTelemetry
```

Nit: We typically write it out in specs

### Histograms

> **Status**: [Stable](../document-status.md)

Reviewer: Can someone confirm this is the current behavior of the prometheus remote write exporter?

If not, then we should probably decouple this change from stabilization, since we can't stabilize something with zero implementations. 

Is there an split the stability level of this section such that we stabilize existing and keep the new in development.

Reviewer 2: I recently did an audit of the PRW exporter, and fixed the discrepancies I found:

* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/47714
* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/47689
* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/47687
* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/47681

So I believe the PRW exporter is fully compliant with the entire specification (excluding possibly the NHCB change proposed here).

### Histograms

> **Status**: [Stable](../document-status.md)

Reviewer 2: ```suggestion
**Status**: [Stable](../document-status.md), except where otherwise specified.
```

> An [OpenTelemetry Histogram](../metrics/data-model.md#histogram) with a cumulative aggregation temporality MUST be converted to a Prometheus Histogram by default, but users may opt-in to converting an OpenTelemetry Histogram to a [Prometheus Native Histogram with Custom Buckets](#histograms-as-prometheus-nhcb) (NHCB) instead when allowed by the Prometheus Protocol.

Reviewer 2: We often tag individual sentences when the reference things we haven't stabilized.
```suggestion
An [OpenTelemetry Histogram](../metrics/data-model.md#histogram) with a cumulative aggregation temporality MUST be converted to a Prometheus Histogram by default.

[Development](../document-status.md): Users may opt-in to converting an OpenTelemetry Histogram to a [Prometheus Native Histogram with Custom Buckets](#histograms-as-prometheus-nhcb) (NHCB) instead when allowed by the Prometheus Protocol.
```

### Histograms

**Status**: [Stable](../document-status.md)

An [OpenTelemetry Histogram](../metrics/data-model.md#histogram) with a cumulative aggregation temporality MUST be converted to a Prometheus Histogram by default, but users may opt-in to converting an OpenTelemetry Histogram to a [Prometheus Native Histogram with Custom Buckets](#histograms-as-prometheus-nhcb) (NHCB) instead when allowed by the Prometheus Protocol.

> OpenTelemetry Histograms with Delta aggregation temporality SHOULD be aggregated into a Cumulative aggregation temporality and follow the logic below, or MUST be dropped.

Reviewer 2: In practice, the delta -> cumulative conversion logic is often separate from the prometheus exporter.  Do you think we should start embedding deltatocumulative converters (or similar) in prometheus collector exporters to satisfy this?

Reviewer 3: I believe we have [similar wording for Sums](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#:~:text=If%20the%20aggregation%20temporality%20is%20delta%20and%20the%20sum%20is%20monotonic%2C%20it%20MAY%20be%20converted%20to%20a%20cumulative%20temporality%20and%20become%20a%20Prometheus%20Counter), with the difference that for sums we use MAY instead of SHOULD.

Personally I'd love to simplify prometheus exporter implementations and just drop, but since we already do something for Sums, I think it doesn't hurt to do the same for histograms

Reviewer 2: I would be in favor of MAY then.

Reviewer: Prometheus SDK exporter already must set temporality to cumualtive: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk_exporters/prometheus.md#temporality

That means they can only get delta from MetricProducers, since the SDK itself will always comply with this request.

I have no interest in requesting that all SDKs perform stateful delta->cumualtive aggregation in their prometheus exporters.

Reviewer 2: Yes, this is only relevant for collector exporters

Reviewer 2: ```suggestion
OpenTelemetry Histograms with Delta aggregation temporality MAY be aggregated into a Cumulative aggregation temporality and follow the logic below, or MUST be dropped.
```

Author: Change to MAY e625399f223e2ba181d7ac656bcfa52bb7693639
